### PR TITLE
Events in Solidity

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

File: libsolidity/AST.h [hunk 1]
```diff
@@ -527,6 +533,37 @@ class ModifierInvocation: public ASTNode
 };
 
 /**
+ * Definition of a (loggable) event.
+ */
+class EventDefinition: public Declaration, public Documented
+{
+public:
+	EventDefinition(Location const& _location,
+					   ASTPointer<ASTString> const& _name,
```
Comment (LefterisJP) [inline libsolidity/AST.h]: Indentation ?

File: libsolidity/ExpressionCompiler.cpp [hunk 2]
```diff
@@ -317,14 +315,42 @@ bool ExpressionCompiler::visit(FunctionCall const& _functionCall)
...
 			{
 				arguments[arg]->accept(*this);
 				appendTypeConversion(*arguments[arg]->getType(), *function.getParameterTypes()[arg], true);
 			}
-			// @todo move this once we actually use memory
-			CompilerUtils(m_context).storeInMemory(0);
-			m_context << u256(32) << u256(0) << eth::logInstruction(logNumber);
+			unsigned length = appendExpressionCopyToMemory(*function.getParameterTypes().front(),
+														   *arguments.front());
+			solAssert(length == 32, "Log data should have length 32.");
```
Comment (LefterisJP) [inline libsolidity/ExpressionCompiler.cpp]: just being pedantic here. Maybe specify "a length of 32 bytes." in the assert message?

File: libsolidity/ExpressionCompiler.cpp [hunk 3]
```diff
@@ -459,14 +485,13 @@ void ExpressionCompiler::endVisit(MemberAccess const& _memberAccess)
...
 {
 	_indexAccess.getBaseExpression().accept(*this);
-	_indexAccess.getIndexExpression().accept(*this);
-	appendTypeConversion(*_indexAccess.getIndexExpression().getType(),
-						 *dynamic_cast<MappingType const&>(*_indexAccess.getBaseExpression().getType()).getKeyType(),
-						 true);
+
+	TypePointer const& keyType = dynamic_cast<MappingType const&>(*_indexAccess.getBaseExpression().getType()).getKeyType();
+	unsigned length = appendExpressionCopyToMemory(*keyType, _indexAccess.getIndexExpression());
+	solAssert(length == 32, "Mapping key length in memory has to be 32.");
```
Comment (LefterisJP) [inline libsolidity/ExpressionCompiler.cpp]: same as in the above assertion ?

File: libsolidity/ExpressionCompiler.h [hunk 4]
```diff
@@ -94,8 +94,13 @@ class ExpressionCompiler: private ASTConstVisitor
 									bool bare = false);
 	/// Appends code that copies the given arguments to memory (with optional offset).
 	/// @returns the number of bytes copied to memory
-	unsigned appendArgumentCopyToMemory(TypePointers const& _functionType, std::vector<ASTPointer<Expression const>> const& _arguments,
-										unsigned _memoryOffset = 0);
+	unsigned appendArgumentCopyToMemory(TypePointers const& _types,
+										 std::vector<ASTPointer<Expression const>> const& _arguments,
```
Comment (LefterisJP) [inline libsolidity/ExpressionCompiler.h]: indentation

**Comment by chriseth** [comment]
Rebased.

**Pull request merged**